FutureNorthants/northants-design-system
# Replace LazyImage with ResponsiveImage

Resolves #390

The react-lazy-progressive-image package is deprecated. This PR replaces it with a custom component called ResponsiveImage. 

- It uses a standard image tag and makes use of the `loading="lazy"` attribute to make the images load when scrolled into view. More information [here](https://developer.mozilla.org/en-US/docs/Web/Performance/Lazy_loading).
- It uses `srcset` to display smaller images on smaller screen resolutions where possible. 
- It allows you to pass in the image ratio, but also allows you to set whether the image should cover the container or be contained within the container. The inspiration came from [here](https://www.smashingmagazine.com/2021/10/object-fit-background-size-css/)
- It has a ratio called 'auto' that allows the image container to fit the parent container. When using this you need to ensure the parent container has `position: relative;`. This is used on the search box, hero and home hero.
- When it is used instead of a background image, the other containers have been updated to use a higher z-index to make the elements sit above the image. I considered using `background-image: image-set()` instead ([more info here](https://developer.mozilla.org/en-US/docs/Web/CSS/image/image-set)), but this is for screen quality, rather than screen resolutions. 

## Testing
- Checkout this branch and run `npm install`
- Run `npm run dev` to manually test the affected components.
- View the Example Content page. Inspect the network tab and scroll down. You should see the images appear when they scroll into view. 
- Run `npm run test` to run the test suite.
